### DISCUSSION TITLE
why one block generate multi bloom filter?

### DISCUSSION THREAD

**Discussion by zdyj3170101136** [grafana/tempo#3059] [Q&A]

**Comment by mapno** [comment]
[Bloom filters](https://en.wikipedia.org/wiki/Bloom_filter) are split into shards, which are each file you see called `bloom-{1,2,...}`. Sharding is used to reduce bloom filter sizes for large blocks, which makes them more cacheable.

The number of shards is the result of calculations over the false positive rate, desired shard size and number of objects in a block. The math behind is not straightforward, but you can check it out in our [code](https://github.com/grafana/tempo/blob/main/tempodb/encoding/common/bloom.go#L25-L52) and the [library](https://github.com/bits-and-blooms/bloom/blob/master/bloom.go#L130-L137) we use.

**Comment by zdyj3170101136** [comment]
> [布隆过滤器](https://en.wikipedia.org/wiki/Bloom_filter)被分成碎片，您看到的每个文件都称为`bloom-{1,2,...}`. 分片用于减少大块的布隆过滤器大小，这使得它们更易于缓存。
> 
> 分片数量是误报率、所需分片大小和块中对象数量的计算结果。[背后的数学并不简单，但您可以在我们的代码](https://github.com/grafana/tempo/blob/main/tempodb/encoding/common/bloom.go#L25-L52)和我们使用的[库](https://github.com/bits-and-blooms/bloom/blob/master/bloom.go#L130-L137)中查看它。

great!

i also have an question.

what is the query speed when using s3 as backend?

(query signle traceid)

**Reply by mapno** [reply]
We don't have official benchmarks. Tempo's read path is highly parallelizable, so it can widely depend on the setup you run. In production deployments, we usually target 1GB/s for baseline search throughput. But to be honest, we don't monitor that closely.
